src/utils/map.ts

```diff
@@ -7,12 +7,27 @@ export module Utils {
    * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map
    */
   export class Map<K, V> {
-    private _keyValuePairs: { key: K; value: V; }[] = [];
+    private _keyValuePairs: { key: K; value: V; }[];
+    private _es6Map: any;
+
+    public constructor() {
+      if (typeof (<any>window).Map === "function") {
+        this._es6Map = new (<any>window).Map();
+      } else {
+        this._keyValuePairs = [];
+      }
+    }
 
     public set(key: K, value: V) {
       if (Utils.Math.isNaN(key)) {
         throw new Error("NaN may not be used as a key to the Map");
```

Comment:
This is ugly. ES6 Map accepts NaN as a valid key. However, we cannot do that because `NaN` is never `===` with `NaN` for the time we do reference equality in the array, hence if we want to fully support the behavior we need to put a hack there to check for NaN